Mathlib/Algebra/Polynomial/Degree/IsMonicOfDegree.lean

```diff
@@ -0,0 +1,269 @@
+/-
+Copyright (c) 2025 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.Algebra.Polynomial.AlgebraMap
+import Mathlib.Algebra.Polynomial.BigOperators
+import Mathlib.Tactic.ComputeDegree
+
+/-!
+# Monic polynomials of given degree
+
+This file defines the predicate `Polynomial.IsMonicOfDegree p n` that states that
+the polynomial `p` is monic and has degree `n` (i.e., `p.natDegree = n`.)
+
+We also provide some basic API.
+-/
+
+namespace Polynomial
+
+/-- This says that `p` has `natDegree` `n` and is monic. -/
+def IsMonicOfDegree {R : Type*} [Semiring R] (p : R[X]) (n : ℕ) : Prop :=
+  p.natDegree = n ∧ p.Monic
+
+lemma IsMonicOfDegree.monic {R : Type*} [Semiring R] {p : R[X]} {n : ℕ}
+    (h : IsMonicOfDegree p n) :
+    p.Monic :=
+  h.2
+
+lemma IsMonicOfDegree.natDegree_eq {R : Type*} [Semiring R] {p : R[X]} {n : ℕ}
+    (h : IsMonicOfDegree p n) :
+    p.natDegree = n :=
+  h.1
+
+lemma IsMonicOfDegree.ne_zero {R : Type*} [Semiring R] [Nontrivial R] {p : R[X]} {n : ℕ}
+    (h : IsMonicOfDegree p n) :
+    p ≠ 0 :=
+  h.2.ne_zero
+
+@[simp]
+lemma isMonicOfDegree_zero {R : Type*} [Semiring R] {p : R[X]} :
+    IsMonicOfDegree p 0 ↔ p = 1 := by
+  simp only [IsMonicOfDegree]
+  refine ⟨fun ⟨H₁, H₂⟩ ↦ eq_one_of_monic_natDegree_zero H₂ H₁, fun H ↦ ?_⟩
+  subst H
+  simp
+
+lemma IsMonicOfDegree.leadingCoeff_eq {R : Type*} [Semiring R] {p : R[X]} {n : ℕ}
+    (hp : IsMonicOfDegree p n) :
+    p.leadingCoeff = 1 :=
+  Monic.def.mp hp.monic
+
+@[simp]
+lemma isMonicOfDegree_iff_of_subsingleton {R : Type*} [Semiring R] [Subsingleton R]
+    (p : R[X]) (n : ℕ) :
+    IsMonicOfDegree p n ↔ n = 0 := by
+  rw [Subsingleton.eq_one p]
+  refine ⟨fun ⟨H, _⟩ ↦ ?_, fun H ↦ ?_⟩
+  · rwa [natDegree_one, eq_comm] at H
+  · rw [H, isMonicOfDegree_zero]
+
+lemma isMonicOfDegree_iff {R : Type*} [Semiring R] [Nontrivial R] (p : R[X]) (n : ℕ) :
+    IsMonicOfDegree p n ↔ p.natDegree ≤ n ∧ p.coeff n = 1 := by
+  simp only [IsMonicOfDegree]
+  refine ⟨fun ⟨H₁, H₂⟩ ↦ ⟨H₁.le, H₁ ▸ Monic.coeff_natDegree H₂⟩, fun ⟨H₁, H₂⟩ ↦ ⟨?_, ?_⟩⟩
+  · exact natDegree_eq_of_le_of_coeff_ne_zero H₁ <| H₂ ▸ one_ne_zero
+  · exact monic_of_natDegree_le_of_coeff_eq_one n H₁ H₂
+
+lemma IsMonicOfDegree.exists_natDegree_lt {R : Type*} [Semiring R] {p : R[X]} {n : ℕ}
```

Comment:
This could potentially be tied in with `eraseLead`, as the "existing" polynomial.

Reply:
OK; I was not aware of `eraseLead`.

Reply:
(This should avoid the import of Mathlib.Algebra.Polynomial.BigOperators.)